Convert IDPF to use auxiliary bus

madhuchittim opened the pull request:
This series converts the idpf driver to use the auxiliary bus.

The main motivation for this patch series is to address the following topics:
1) The auxiliary bus is preferred in the Linux community.
2) Upstreaming of dynamic vports and ADI functionality requires driver 
   separation into PCIe and Ethernet driver component.
3) IDPF PCIe driver supports multiple sub functions like RDMA, ethernet etc.

First 5 patches are in preparation for enabling the code changes for auxiliary
bus plug in. They are only refactoring the existing code to different files. The
similar functionality is accomodated into different files. This enables separating
the ethernet functionality for future patches.

Patch 6 introduces the IDC shared header file. This file is for inter device
communication.

Patch 7 creates logical separation of ethernet portion of the code, but it is
still a global single adapter for all the vports.

Patch 8 makes it multi-instance driver like each vport will have its own adapter instance.
Patch 9 moves the ethernet functionality into eth directory.
And finally Patch 10 connect the ethernet portion of the driver to the auxiliary bus.

Commit:
idpf: Refactor filter operations in separate file

Added new files 'idpf_fltr.c' and 'idpf_fltr.h'.

This patch makes an effort to move all filter operations into a
separate file. This reduces the burden on other files where these
operations were clubbed together, and makes them look clean now.

Also, moved some of the structures from idpf_virtchnl.c to
idpf_virtchnl.h

There is no change in the code or functionality before and after
the change.

Signed-off-by: Tarun K Singh <tarun.k.singh@intel.com>

Commit:
idpf: Refactor netdev operations in new file

Added new files 'idpf_netdev.c' and 'idpf_netdev.h'.

This patch moves all netdev operations from 'idpf_lib.c' into new
file 'idpf_netdev.c'. It makes them look clean now.

There is no change in the code or functionality before and after the
change.

Signed-off-by: Tarun K Singh <tarun.k.singh@intel.com>

Commit:
idpf: Refactor vport operations in new file

Added new files 'idpf_vport.c' and 'idpf_vport.h'.

This patch moves vport related functions in new file 'idpf_vport.c'.
The header related definitions are moved to new file 'idpf_vport.h'.
Again the purpose is to separate the vport functionality into its
own file.

There is no change in the code or functionality before and after
the change.

Signed-off-by: Tarun K Singh <tarun.k.singh@intel.com>

Commit:
idpf: Refactor vport config operations in new file

Added new file 'idpf_vport_config.c'.

It moves vport config functions in new file 'idpf_vport_config.c'.
The purpose is to separate the vport config functionality into
separate file.

There is no change in the code or functionality before and after
the change.

Signed-off-by: Tarun K Singh <tarun.k.singh@intel.com>

Commit:
idpf: Refactor vport q transactions in new file

Added new file 'idpf_vport_q.c'.

This patch moves vport q transactions in new file 'idpf_vport_q.c'.

There is no change in the code or functionality before and after
the change.

Signed-off-by: Tarun K Singh <tarun.k.singh@intel.com>

Commit:
idpf: Add Inter-Device Communication interface file

Added new file 'idpf_eth_idc.h'.

Added idpf_eth_shared as common information to share between main
and ethernet device. Moved a member from idpf_adapter in new
eth shared structure, and moved some other structs: idpf_flags,
idpf_vector_info, idpf_intr_reg, idpf_q_vector into new file
'idpf_eth_idc.h' to lay foundation work for inter device
communication (idc) between main and ethernet device.

There is no change in the functionality before and after the change.

Signed-off-by: Singh, Tarun K <tarun.k.singh@intel.com>
Signed-off-by: Shailendra Bhatnagar <shailendra.bhatnagar@intel.com>

DanNowlin committed:
ice: Fix package download algorithm

Previously, the driver assumed that all signature segments would contain
one or more buffers to download. In the future, there will be signature
segments that will contain no buffers to download.

Correct download flow to allow for signature segments that have zero
download buffers and skip the download in this case.

Fixes: 3cbdb0343022 ("ice: Add support for E830 DDP package segment")
Reviewed-by: Przemek Kitszel <przemyslaw.kitszel@intel.com>
Signed-off-by: Dan Nowlin <dan.nowlin@intel.com>
Signed-off-by: Paul Greenwalt <paul.greenwalt@intel.com>
Reviewed-by: Paul Menzel <pmenzel@molgen.mpg.de>

pgreenwa committed:
ice: add additional E830 device ids

Add support for additional E830 device ids which are supported by the
driver:
- 0x12D5: Intel(R) Ethernet Controller E830-C for backplane
- 0x12D8: Intel(R) Ethernet Controller E830-C for QSFP
- 0x12DA: Intel(R) Ethernet Controller E830-C for SFP
- 0x12DC: Intel(R) Ethernet Controller E830-XXV for backplane
- 0x12DD: Intel(R) Ethernet Controller E830-XXV for QSFP
- 0x12DE: Intel(R) Ethernet Controller E830-XXV for SFP

Reviewed-by: Wojciech Drewek <wojciech.drewek@intel.com>
Signed-off-by: Paul Greenwalt <paul.greenwalt@intel.com>

pgreenwa committed:
ice: update E830 device ids and comments

Update existing E830 device ids and comments to align with new naming 'C'
for 100G and 'CC' for 200G.

Reviewed-by: Wojciech Drewek <wojciech.drewek@intel.com>
Signed-off-by: Paul Greenwalt <paul.greenwalt@intel.com>

Commit:
idpf: Connect ethernet part to the auxiliary bus

It now uses the auxiliary bus to manage ethernet part of driver and
its devices.

In 'struct idpf_eth_idc_auxiliary_driver' added new member to save
signature, and another member 'struct auxiliary_driver adrv' to keep
auxiliary device information. Renamed function idpf_eth_device_add()
to idpf_eth_probe() to reflect auxiliary device probe.

Added new function 'idpf_eth_idc_driver_register()' to register
auxiliary driver. And added corespondig call to unregister the
auxiliary driver in function 'idpf_eth_idc_driver_unregister()'.

Added module_init(idpf_driver_register), and
module_exit(idpf_driver_unregister) to register/unregister the
auxiliary device information.

Added management auxiliary call 'auxiliary_device_init()' and
'auxiliary_device_add()' to initialize the device and to request the
system to start the probe for device(s).

Added new function 'idpf_eth_idc_device_deinit()' to manage the
delete and uninit the auxuliary devices.

Removed unused function after conversion idpf_eth_unregister(), and
removed variables idpf_context, and eth_context as they are not
required. In this new model those are retrieved from the system
devices references now on need basis.

Signed-off-by: Tarun K Singh <tarun.k.singh@intel.com>
Signed-off-by: Shailendra Bhatnagar <shailendra.bhatnagar@intel.com>

jason77-wang committed:
e1000e: move force SMBUS near the end of enable_ulp function

The commit 861e8086029e ("e1000e: move force SMBUS from enable ulp
function to avoid PHY loss issue") introduces a regression on
CH_MTP_I219_LM18 (PCIID: 0x8086550A). Without this commit, the
ethernet works well after suspend and resume, but after applying the
commit, the ethernet couldn't work anymore after the resume and the
dmesg shows that the NIC Link changes to 10Mbps (1000Mbps originally):
[   43.305084] e1000e 0000:00:1f.6 enp0s31f6: NIC Link is Up 10 Mbps Full Duplex, Flow Control: Rx/Tx

Without the commit, the force SMBUS code will not be executed if
"return 0" or "goto out" is executed in the enable_ulp(), and in my
case, the "goto out" is executed since FWSM_FW_VALID is set. But after
applying the commit, the force SMBUS code will be ran unconditionally.

Here move the force SMBUS code back to enable_ulp() and put it
immediate ahead of hw->phy.ops.release(hw), this could allow the
longest settling time as possible for interface in this function and
doesn't change the original code logic.

Fixes: 861e8086029e ("e1000e: move force SMBUS from enable ulp function to avoid PHY loss issue")
Signed-off-by: Hui Wang <hui.wang@canonical.com>
Acked-by: Vitaly Lifshits <vitaly.lifshits@intel.com>

Commit:
ice: Deduplicate tc action setup

ice_tc_setup_redirect_action() and ice_tc_setup_mirror_action() are almost
identical, except for setting filter action. Reduce them to one function
with an extra param, which handles both cases.

Reviewed-by: Mateusz Polchlopek <mateusz.polchlopek@intel.com>
Signed-off-by: Marcin Szycik <marcin.szycik@linux.intel.com>
Reviewed-by: Simon Horman <horms@kernel.org>

Commit:
i40e: flower: validate control flags

This driver currently doesn't support any control flags.

Use flow_rule_has_control_flags() to check for control flags,
such as can be set through `tc flower ... ip_flags frag`.

In case any control flags are masked, flow_rule_has_control_flags()
sets a NL extended error message, and we return -EOPNOTSUPP.

Only compile-tested.

Signed-off-by: Asbjørn Sloth Tønnesen <ast@fiberby.net>
Reviewed-by: Simon Horman <horms@kernel.org>

Commit:
iavf: flower: validate control flags

This driver currently doesn't support any control flags.

Use flow_rule_has_control_flags() to check for control flags,
such as can be set through `tc flower ... ip_flags frag`.

In case any control flags are masked, flow_rule_has_control_flags()
sets a NL extended error message, and we return -EOPNOTSUPP.

Only compile-tested.

Signed-off-by: Asbjørn Sloth Tønnesen <ast@fiberby.net>
Reviewed-by: Simon Horman <horms@kernel.org>

Commit:
ice: flower: validate control flags

This driver currently doesn't support any control flags.

Use flow_rule_has_control_flags() to check for control flags,
such as can be set through `tc flower ... ip_flags frag`.

In case any control flags are masked, flow_rule_has_control_flags()
sets a NL extended error message, and we return -EOPNOTSUPP.

Only compile-tested.

Signed-off-by: Asbjørn Sloth Tønnesen <ast@fiberby.net>
Reviewed-by: Michal Swiatkowski <michal.swiatkowski@linux.intel.com>

Commit:
igb: flower: validate control flags

This driver currently doesn't support any control flags.

Use flow_rule_match_has_control_flags() to check for control flags,
such as can be set through `tc flower ... ip_flags frag`.

In case any control flags are masked, flow_rule_match_has_control_flags()
sets a NL extended error message, and we return -EOPNOTSUPP.

Only compile-tested.

Signed-off-by: Asbjørn Sloth Tønnesen <ast@fiberby.net>
Reviewed-by: Simon Horman <horms@kernel.org>

vlifshts committed:
e1000e: change usleep_range to udelay in PHY mdic access

This is a partial revert of commit 6dbdd4de0362 ("e1000e: Workaround
for sporadic MDI error on Meteor Lake systems"). The referenced commit
introduced an issue on vPro systems, where disconnecting and reconnecting
the LAN cable might result in a kernel panic.

This was root caused to the usage of usleep_range in an atomic content
while trying to access the PHY. Change back the usleep_range calls to
udelay.

Fixes: 6dbdd4de0362 ("e1000e: Workaround for sporadic MDI error on Meteor Lake systems")
Co-developed-by: Sasha Neftin <sasha.neftin@intel.com>
Signed-off-by: Sasha Neftin <sasha.neftin@intel.com>
Signed-off-by: Vitaly Lifshits <vitaly.lifshits@intel.com>

walking-machine committed:
ice: Interpret .set_channels() input differently

A bug occurs because a safety check guarding AF_XDP-related queues in
ethnl_set_channels(), does not trigger. This happens, because kernel and
ice driver interpret the ethtool command differently.

How the bug occurs:
1. ethtool -l <IFNAME> -> combined: 40
2. Attach AF_XDP to queue 30
3. ethtool -L <IFNAME> rx 15 tx 15
   combined number is not specified, so command becomes {rx_count = 15,
   tx_count = 15, combined_count = 40}.
4. ethnl_set_channels checks, if there are any AF_XDP of queues from the
   new (combined_count + rx_count) to the old one, so from 55 to 40, check
   does not trigger.
5. ice interprets `rx 15 tx 15` as 15 combined channels and deletes the
   queue that AF_XDP is attached to.

Interpret the command in a way that is more consistent with ethtool
manual [0] (--show-channels and --set-channels).

Considering that in the ice driver only the difference between RX and TX
queues forms dedicated channels, change the correct way to set number of
channels to:

ethtool -L <IFNAME> combined 10 /* For symmetric queues */
ethtool -L <IFNAME> combined 8 tx 2 rx 0 /* For asymmetric queues */

[0] https://man7.org/linux/man-pages/man8/ethtool.8.html

Fixes: 87324e747fde ("ice: Implement ethtool ops for channels")
Reviewed-by: Michal Swiatkowski <michal.swiatkowski@linux.intel.com>
Signed-off-by: Larysa Zaremba <larysa.zaremba@intel.com>

Commit:
idpf: Add ethernet device logical separation

Added new files 'idpf_eth_idc.c', idpf_eth.c, idpf_eth.h, and
'idpf_eth_common.h'

Added new 'struct idpf_eth_adapter' to make ethernet device's
logical separation from main driver. Moved all ethernet related
members into this new structure, and corresponding functions
into new files 'idpf_eth.c' and 'idpf_eth.h'. Added inter device
communication (idc) functions in new file 'idpf_eth_idc.c'. The
idc file contains all the necessary hooks to communicate with
separated logical device (eth). Many ethernet device code function
argumets are now changed from idpf_adapter to idpf_eth_adapter
in corresponding ethernet code files.

Added 'struct idpf_eth_idc_auxiliary_dev' for tracking eth device
creation and deletion. The eth device information is stored in
idpf_eth_shared structure for easy access in both lower and upper
layers.

A new 'struct idpf_eth_idc_dev_info' is added to pass device init
parameters like queue, capabilities etc. to ethernet device from
main driver. The capabilities (caps) macros are moved into the
file 'idpf_eth_idc.h', so that both main and ethernet drivers can
access these macros.

After the primary pci device's reset and initialization flow, a
new task 'idc_eth_init_task' is scheduled. This new task creates
ethernet device structure information, allocates ethernet adapter
structure, and initializes eth portion of the driver which now
takes over the processing of vports of ethernet device (eth).

This patch impliments ethernet device as a single ethernet device
adapter having default single or multiple vport(s).

Ethernet device initialization is a two part process through
functions idpf_eth_device_pre_init(), and idpf_eth_device_post_init().
Pre init runs task idpf_eth_device_pre_init_task() which
checks to see if all vports devices are pre-initialized and ready
for moving on to post init step. Post init step enables the vport's
availability into the system.

Added events for propagating the information from main driver to
ethernet portion of the driver.Added 'enum idpf_eth_idc_event_type'
to indicate whether an event is for all vports/ethernet adapters or
single vport/ethernet adapter.

Added a local flags declaration into 'struct idpf_eth_adapter', and
corresponding 'enum idpf_eth_flags' for keeping lower level events
information.

Removed reference or dependency of 'struct idpf_adapter' i.e. the
header file 'idpf.h' from all the files related to ethernet
operations. These files are: idpf_eth.c/h, idpf_mac.c/h,
idpf_netdev.c/h, idpf_vport.c/h, idpf_vport_config.c,
idpf_vport_q.c, idpf_txrx.c, and idpf_singleq_txrx.c. These files
now include idpf_eth.h instead of idpf.h.

Renamed variable 'vport_ctrl_lock' to 'reset_lock' and associated
function to reflect its usage nature. Removed main driver's lock
dependency into the ethernet drivers. Ethernet driver code now
refers to local version og 'vport_ctrl_lock' and flags in
eth_adapter to take decisions. The local flags information is
updated via events from main driver.

In function idpf_decfg_netdev(), added a netdev register check
before unregistering.

The content of 'idpf_virtchnl.h' are moved into 'idpf_eth_common.h'
as they are common for both, and file 'idpf_virtchnl.h' has been
removed.

Signed-off-by: Tarun K Singh <tarun.k.singh@intel.com>
Signed-off-by: Shailendra Bhatnagar <shailendra.bhatnagar@intel.com>

Commit:
idpf: Make ethernet device a multi-instance device

Added capability to allocate multiple adapter instance based on
number of vports. Most of the double pointers into 'struct
idpf_eth_adapter'are converted into single pointer to hold data
structure for vport, netdev, vport_config, vport_params_reqd, and
vport_params_recvd.

Removed unused following function(s) while making these changes:
idpf_get_free_slot(), idpf_vport_params_buf_alloc(),
idpf_vid_to_vport(), and few others.

In 'struct idpf_eth_adapter' removed pre_init_task, pre_init_wq,
members vport_ids, next_vport, and num_alloc_vports. They are not
required into the multi instance adapter implementation.

In 'struct idpf_eth_idc_dev_info' a new member
'bool start_probe_post_init' is added so that the probe post init
can be triggered concurrently. Added new event to trigger the post
initialization part of ethernet probe.

Added new function idpf_eth_idc_vid_to_dev_info() to retrieve the
device instance based on vport id associated with it.

Signed-off-by: Tarun K Singh <tarun.k.singh@intel.com>
Signed-off-by: Shailendra Bhatnagar <shailendra.bhatnagar@intel.com>

Commit:
idpf: Separate ethernet files in new directory eth

Created new directory 'eth' to separate the ethernet driver files.

Following files are now part of 'eth' and represents ethernet
portion of the driver.
Files:idpf_txrx.c, idpf_txrx.h, idpf_singleq_txrx.c, idpf_ethtool.c,
idpf_mac.c, idpf_mac.h, idpf_netdev.c, idpf_netdev.h, idpf_vport.c
idpf_vport.h, idpf_vport_config.c, idpf_vport_q.c, idpf_eth.c, and
idpf_eth.h

Makefile modification are done to include directory name 'eth' for
above moved 'c' files, and a flag.

Signed-off-by: Tarun K Singh <tarun.k.singh@intel.com>

halfboy93 committed:
ice: refactor struct ice_vsi_cfg_params to be inside of struct ice_vsi

Refactor struct ice_vsi_cfg_params to be embedded into struct ice_vsi.
Prior to that the members of the struct were scattered around ice_vsi,
and were copy-pasted for purposes of reinit.
Now we have struct handy, and it is easier to have something sticky
in the flags field.

Suggested-by: Przemek Kitszel <przemyslaw.kitszel@intel.com>
Reviewed-by: Przemek Kitszel <przemyslaw.kitszel@intel.com>
Reviewed-by: Vaishnavi Tipireddy <vaishnavi.tipireddy@intel.com>
Signed-off-by: Mateusz Polchlopek <mateusz.polchlopek@intel.com>

pgreenwa committed:
ice: fix 200G PHY types to link speed mapping

Commit 24407a01e57c ("ice: Add 200G speed/phy type use") added support
for 200G PHY speeds, but did not include the mapping of 200G PHY types
to link speed. As a result the driver is returning UNKNOWN link speed
when setting 200G ethtool advertised link modes.

To fix this add 200G PHY types to link speed mapping to
ice_get_link_speed_based_on_phy_type().

Fixes: 24407a01e57c ("ice: Add 200G speed/phy type use")
Reviewed-by: Michal Swiatkowski <michal.swiatkowski@linux.intel.com>
Signed-off-by: Paul Greenwalt <paul.greenwalt@intel.com>

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,374 @@
+		idpf_send_get_stats_msg(vport);
+
+	queue_delayed_work(adapter->stats_wq, &adapter->stats_task,
+			   msecs_to_jiffies(10000));
```
i don't think its 10000. did you check base code

madhuchittim replied:
Verified in upstream code base and it is aligned to it.

etantilov commented on drivers/net/ethernet/intel/idpf/idpf.h:
```diff
@@ -244,51 +227,31 @@ struct idpf_adapter {
-	bool req_rx_splitq;
 
-	struct mutex vport_ctrl_lock;
+	struct mutex reset_lock;
```
The function of vport_ctrl_lock was to protect instances where the vport data might be lost. For example if ethtool ndo arrives during the reset. From what I can see reset_lock is only taken in idpf_init_hard_reset() and not checked anywhere else, so I am not sure how we are using it to protect access during reset.

tksingh108 replied:
The reset information is propagated to the aux drivers instance via event interface from PCI - IDC- eth. And it uses those flags for protection now.

etantilov replied:
The RESET_IN_PROGRESS flag is not sufficient though, since it does not guarantee that vport data will be in good state. The use of flags was specifically rejected by the maintainer - the addition of vport_ctrl_lock was the last change that got us in. So this code aside from not actually protecting anything (just test with any ndo during reset) is not in a format that will be accepted by netdev. We need to figure out some mechanism for robust locking between the instances.

tksingh108 replied:
I am trying to get over this code again to see if I can reapply the similar lock. I will update back on it.

tksingh108 replied:
I think I addressed this issue finally. The code changes are in parity with the previous code now. The similar locking mechanism has been applied to the aux driver now.

tksingh108 replied:
I hope we are okay with the changes you requested. Please let me know if we still have concern about it. I am marking it as resolved.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+
+	/* Following requests are from auxiliary eth driver to main driver */
+	IDPF_ETH_IDC_EVENT_REQ_HARD_RESET,
+	IDPF_ETH_IDC_EVENT_ENTER_PROMISCUOUS_MODE,
```
PCI driver should not deal with any network specific functionality. So not sure why this PROMISCUOUS event is required??

tksingh108 replied:
addressed in current changes.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+ * the event type IDPF_ETH_IDC_EVENT_SINGLE_VPORT is routed to only the
+ * intended port.
+ */
+enum idpf_eth_idc_event_type {
```
Every eth aux dev will represent one vport. So these events purpose is not very clear.

tksingh108 replied:
There are two type of events:
events routed to all the vports, and event routed to single vport. We need a mechanism to route the events differently. This serves this purpose. Let me know if it is clear now or not?

tksingh108 replied:
I hope we are ok with the current patch change push.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+#define _IDPF_ETH_IDC_H_
+
+#include <linux/auxiliary_bus.h>
+#include "idpf_eth_common.h"
```
This header should be independent as much as possible. what is the common header dependency??

tksingh108 replied:
Done ! Moved the include file to idpf.h file.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+struct idpf_adapter;
+struct idpf_vc_xn_params;
+
+#define IDPF_MAX_Q				16
```
These defines are not standard. They are pci driver specific resource management. So should be moved out of this header. They can be moved to eth_common.h i assume.

tksingh108 replied:
Done!

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+	IDPF_ETH_IDC_EVENT_LINK_CHANGE,
+	IDPF_ETH_IDC_EVENT_RESET_INITIATED,
+	IDPF_ETH_IDC_EVENT_RESET_COMPLETE,
+	IDPF_ETH_IDC_EVENT_DEINIT,
```
DEINIT and REMOVE seems to stand out in this list of events. We need to discuss a bit deep on why these are required. Atleast when the aux device is deleted, the auxdriver remove should be called by the aux bus(not by the pci driver)

tksingh108 replied:
Remove is not present anymore in current changes. The other Deinit event is still there and probably needed based on the design of the code we have at the moment. May be this can be re-worked later as a separate PR to get rid of this event because it may alter the existing code and flow which is not in best interest of this PR.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+	struct idpf_eth_shared *eth_shared;
+	enum idpf_vport_type vport_type;
+	struct idpf_eth_idc_auxiliary_dev_caps caps;
+	bool start_probe_post_init;
```
??aux probe is called by aux bus right??this is not clear...

bhatnagarintel replied:
aux probe is the side effect of calling auxiliary_device_add. The device add cannot be performed again during reset. Hence the need for explicit invocation.

tksingh108 replied:
There is also a workaround in the driver to run the pre-init portion of the code sequentially which can be done only by controlling how we initialize the aux driver. When card's firmware supports parallel execution, we can remove this dependency.

tksingh108 replied:
Moved to the eth adapter.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+	enum idpf_vport_type vport_type;
+	struct idpf_eth_idc_auxiliary_dev_caps caps;
+	bool start_probe_post_init;
+	u32 vport_id;
```
This data is passed by pci driver to aux driver. And aux driver is the one which sends create_ vport. So not sure how a pci driver can have this data

tksingh108 replied:
As we talked about it, there are case when the event such as link up or down, the PCI driver receives notification of such events and its device identifier which is port id in our case. This is used to locate the correct eth adapter so that the event can be routed to the correct aux device.

tksingh108 replied:
I think we agreed to keep it this way for now. Marking it resolved.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+	((struct idpf_adapter *)((dev_info)->idpf_context))
+#define idpf_eth_shared_to_dev(eth_shared) (&(eth_shared)->pdev->dev)
+
+int idpf_eth_idc_init_shared(struct idpf_eth_shared *eth_shared,
```
This is pci driver specific code. Only the initialized sturct is passed in IDC. So should be moved out of this header.

tksingh108 replied:
Done!

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+			     struct pci_dev *pdev);
+void idpf_eth_idc_deinit_shared(struct idpf_eth_shared *eth_shared);
+
+void idpf_eth_idc_device_init(struct idpf_adapter *adapter);
```
Same as above

tksingh108 replied:
done !

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+void idpf_eth_idc_device_init(struct idpf_adapter *adapter);
+void idpf_eth_idc_device_deinit(struct idpf_adapter *adapter);
+
+int idpf_eth_idc_driver_register(void);
```
i think this also belongs to pci driver..... so should be moved out of this header

tksingh108 replied:
done !

nmkwan committed:
ice: Do not get coalesce settings while in reset

Getting coalesce settings while reset is in progress can cause NULL
pointer deference bug.
If under reset, abort get coalesce for ethtool.

Fixes: 67fe64d78c43 ("ice: Implement getting and setting ethtool coalesce")
Signed-off-by: Ngai-Mint Kwan <ngai-mint.kwan@intel.com>
Reviewed-by: Mateusz Polchlopek <mateusz.polchlopek@intel.com>
Signed-off-by: Pawel Chmielewski <pawel.chmielewski@intel.com>
Reviewed-by: Simon Horman <horms@kernel.org>
Signed-off-by: Dawid Osuchowski <dawid.osuchowski@linux.intel.com>

Commit:
ice: remove correct filters during eswitch release

ice_clear_dflt_vsi() is only removing default rule. Both default RX and
TX rule should be removed during release.

If it isn't switching to switchdev, second time results in error, because
TX filter is already there.

Fix it by removing the correct set of rules.

Fixes: 50d62022f455 ("ice: default Tx rule instead of to queue")
Reviewed-by: Wojciech Drewek <wojciech.drewek@intel.com>
Signed-off-by: Michal Swiatkowski <michal.swiatkowski@linux.intel.com>
Signed-off-by: Marcin Szycik <marcin.szycik@linux.intel.com>
Reviewed-by: Simon Horman <horms@kernel.org>

ij-intel committed:
net: e1000e & ixgbe: Remove PCI_HEADER_TYPE_MFD duplicates

PCI_HEADER_TYPE_MULTIFUNC is define by e1000e and ixgbe and both are
unused. There is already PCI_HEADER_TYPE_MFD in pci_regs.h anyway which
should be used instead so remove the duplicated defines of it.

Signed-off-by: Ilpo Järvinen <ilpo.jarvinen@linux.intel.com>
Reviewed-by: Hariprasad Kelam <hkelam@marvell.com>
Reviewed-by: Jesse Brandeburg <jesse.brandeburg@intel.com>

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.c:
```diff
@@ -0,0 +1,453 @@
+			if (!eth_dev)
+				continue;
+
+			eth_idc_drv->adrv.probe(&eth_dev->adev, NULL);
```
why is pci driver calling the aux driver probe directly? aux bus should take care of the probe for all its devices all the time. In the reset flow, aux device is present and the probe is already called. Aux driver should receive the reset info through an event and it has to readjust itself one level below its probe.(no difference here like what PCI device/driver does)

tksingh108 replied:
This is to run the initialization in controlled fashion. We have restriction that we cannot run the init in parallel for all the vports.
If firmware supports parallel execution of vport creation, we can make modifications to the init logic.

burra006 replied:
This will be very difficult to get upstream acceptance i assume, as it changes the very basic flows of bus/driver model. Better we raise the concern to HMA/CP and ask them to prioritize the fix.

Another confusion atleast for me is, you are doing it only in reset flows. Not in driver init flows???So if it is a problem, then it has to exist in both cases?

tksingh108 replied:
Init flow also follows reset path. The first initialization starts after reset, and mail box init commands etc, so it follows the same path for first time initialization. That's how it has been coded.
I will check if I can remove calling direct probe from PCI driver for successive resets etc. by making some code changes.

bhatnagarintel replied:
If hard reset needs to always do probe indirectly (by calling aux dev add), you will have to free the full aux dev structure which houses pointers to the 'eth' adapter and vport_config. 
So we would need to cache vport_config elsewhere (and whatever else in the eth adapter is needed across resets)

tksingh108 replied:
I have removed the dependency of PCI driver calling probe directly. It now only calls once to initialize/start the device. Later auxiliary device handles the probe by itself. The code in contained in the aux driver. Only the serialization part we have dependency on the PCI driver. Please check the changes, and let me know if I can resolve this.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.c:
```diff
@@ -0,0 +1,453 @@
+	enum idpf_queue_type queue_type;
+	u16 i, err;
+
+	for (i = 0; i < adapter->default_vports; ++i) {
```
we entered this function in a for loop of number of default_vports. So another for loop of same does not seems to be reasonable. if you want this loop for finding the empty slot, then run a while loop on adevs.

tksingh108 replied:
Removed extra looping.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.c:
```diff
@@ -0,0 +1,453 @@
+		new_eth_dev->eth_info.vport_type = vport_type;
+
+		if (vport_type == IDPF_DEFAULT_VPORT)
+			queue_type = IDPF_QUEUE_DEFAULT_VPORT;
```
why do we need another enum? can we reuse the IDPF_DEFAULT_VPORT to differentiate the number of queues to be allocated?

tksingh108 replied:
Removed the duplicate definition.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.c:
```diff
@@ -0,0 +1,453 @@
+						    IDPF_DEFAULT_VPORT);
+		if (!eth_dev)
+			return;
+		eth_dev->adev.name = "eth";
```
These three variables can also be moved to idc_device_alloc function to have all the initialization at one place. Will look cleaner.

tksingh108 replied:
Done!

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.c:
```diff
@@ -0,0 +1,453 @@
+	eth_dev->eth_info.caps.crc_enable = adapter->crc_enable;
+
+	/* Queue Info */
+	eth_dev->eth_info.caps.q_info.max_rxq =
```
This is taken care as part of alloc_max_qs in the calling function right??

tksingh108 replied:
Initially PCI driver only passing those max values to aux device. IDC then pre-allocate for the aux device.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.c:
```diff
@@ -0,0 +1,453 @@
+
+extern struct idpf_eth_idc_auxiliary_driver *idpf_eth_get_driver(void);
+/**
+ * idpf_eth_idc_get_driver - returns eth driver ops
```
please check all the comment headers. Many places there is a mismatch...

tksingh108 replied:
Let me know if I missed something. Needs review.

madhuchittim commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,380 @@
+ * idpf_eth_shutdown - Shutdown call
+ * @adev: Structure related to auxiliary ethernet device
+ */
+static void idpf_eth_shutdown(struct auxiliary_device *adev)
```
Can we remove this empty function?

tksingh108 replied:
Done!

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,380 @@
+}
+
+/**
+ * idpf_eth_device_pre_init - Ethernet device pre-initialization routine
```
If possible avoid the pre-init and post-init terminology. If there is such a thing as pre-init, it would be in main driver and as part of the "pre-init" it is allocating aux device, core eth dev info and initializing these 2 structures. Thats how I would look at "pre-init". So there is nothing like post-init - its just probe.  Its possible though that I have not understood the rationale. So please elaborate.

tksingh108 replied:
It is pre-init and post init portion of the same probe under auxiliary device. This whole code belongs to the auxiliary driver, so it shouldn't create any confusion.
Basically aux probe is spilit into two.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,380 @@
+
+	if (adapter->netdev) {
+		netdev = adapter->netdev;
+		if (netdev) {
```
You just checked netdev above. No need to check null again

tksingh108 replied:
Removed extra check!

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_netdev.h:
```diff
@@ -0,0 +1,82 @@
+ * @netstats: Packet and byte stats
+ * @stats_lock: Lock to protect stats update
+ */
+struct idpf_netdev_priv {
```
You can keep one of the two here : vport or adapter. Both should not be needed.
From adapter you can reach vport   adapter->vport
From vport you can reach adapter    vport->adapter

tksingh108 replied:
This is the old structure, I did not add this in this PR. I think this should be done in separate PR if needed.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+	int (*intr_reg_init)(struct idpf_eth_idc_dev_info *dev_info,
+			     u16 num_vecs, struct idpf_q_vector *q_vectors,
+			      u16 *q_vector_idxs);
+	int (*intr_init_vec_idx)(struct idpf_eth_idc_dev_info *dev_info,
```
This would be direct function call as per past comments during review. Atleast thats how  I have it in OOT .
The function of course resides in main driver

tksingh108 replied:
I think we have settled all this. All the PCI calls are direct calls from IDC, these are auxiliary calls only function pointer from IDC to aux etc.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+				 u16 num_vecs,
+				 struct idpf_q_vector *q_vectors,
+				 u16 *q_vector_idxs);
+	int (*req_rel_vec_idx)(struct idpf_eth_idc_dev_info *dev_info,
```
Again direct function call in OOT PR .

tksingh108 replied:
I think we agreed on this IDC interface now. IDC file calls the PCI functions directly, and we have function pointers to aux  files from IDC file.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_vport_config.c:
```diff
@@ -0,0 +1,1714 @@
+	}
+
+do_memcpy:
+	memcpy(rss_data->rss_key, recv_rk->key_flex, rss_data->rss_key_size);
```
recv_rk memory is leaked here unless I am missing something.

tksingh108 replied:
Not sure what leak is referred here. The driver loads fine and I didn't see any issue with the driver.

bhatnagarintel replied:
I referred specifically to recv_rk. If you examine the function, I don't see it being freed. Loading of driver has nothng to do with a few memory blocks leaking.

tksingh108 replied:
I will keep the original code as it as that is not the purpose of this PR . The other issues (non aux related changes) can be addressed with a separate PR.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,380 @@
+
+	switch (event->event_code) {
+	case IDPF_ETH_IDC_EVENT_RESET_INITIATED:
+		set_bit(IDPF_ETH_RESET_IN_PROG, adapter->flags);
```
just bring this set_bit inside the mutex lock block

tksingh108 replied:
done !

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,380 @@
+ * idpf_eth_get_driver - returns ethernet device driver structure
+ * @void: void
+ */
+struct idpf_eth_idc_auxiliary_driver *idpf_eth_get_driver(void)
```
This func is not needed. Please check OOT PR. As per review comments, other means are used to get aux driver pointer in the main driver.

tksingh108 replied:
Please look at the function "idpf_eth_idc_driver_register". It is needed by this function, it doesn't pass any argument. Since it needs it this way, I don't want to keep two version of it. I think it is more clean. Why does it matter how we access this pointer?

bhatnagarintel replied:
Yea and the one version which was recommended in past reviews was to follow the RDMA /IDC model. So I changed it to that. Please check the OOT PR but I am pasting relevant snippet below.

static inline struct idpf_eth_auxiliary_driver *
idpf_eth_dev_to_driver(struct idpf_idc_eth_auxiliary_dev *eth_dev)
{
        return container_of(eth_dev->adev.dev.driver,
                            struct idpf_eth_auxiliary_driver,
                            adrv.driver);
}

tksingh108 replied:
Please see my earlier comment. My version of implementation doesn't have argument to this function. It is void. So it is not possible for me to use container.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/Makefile:
```diff
@@ -1,18 +1,29 @@
 # SPDX-License-Identifier: GPL-2.0-only
-# Copyright (C) 2023 Intel Corporation
+# Copyright (C) 2024 Intel Corporation
```
Copyright year is year of coding, not year of modification.

madhuchittim replied:
Pavan had a comment on this to be changed. @plinga1 Can you please confirm on this.

plinga1 replied:
There was a comment when we upstreamed IDPF to update the copyright year to the current year the driver makes into the upstream. Couldnt find the comment but you may consult Tony for additional info.

bhatnagarintel replied:
See here is my past experience with Copyright. When the file is first coded, the copyright line shows that year e.g.
Copyright (C) 2021 Rafał Miłecki <rafal@milecki.pl>

Now suppose above file is modified again in 2022,  the copyright should be 
Copyright (C) 2021-2022 Rafał Miłecki <rafal@milecki.pl>

Then the file gets modified in 2024 (means not modified in 2023), the copyright should read:
Copyright (C) 2021-2022, 2024 Rafał Miłecki <rafal@milecki.pl>

Essentially, every year that you modify the file should be listed via range (using dash) or explicitly

tksingh108 replied:
I initially talked to Tony about it, and I had left it to the original code date of this driver based on the feedback from him. I will ask him again about it.

madhuchittim replied:
I have caused a problem here.. some of the files I moved the year based on Pavan's comments so some of them would end up showing as 2024

tksingh108 replied:
No worry, I will fix them again.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,380 @@
+	dev_info = adapter->dev_info;
+	eth_shared = idpf_eth_adapter_shared(adapter);
+	pdev = eth_shared->pdev;
+	adapter->req_tx_splitq = true;
```
Should we remove these flags if they are always true ? Or is there some mechanism to make it false

tksingh108 replied:
It is part of original code. I don't want to divert from the definition of this PR for making further changes to this PR.

tksingh108 replied:
How the OOT is doing? I am not sure why it is there in first place? Will it ever need in future?

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,380 @@
+
+	eth_dev = container_of(adev, struct idpf_eth_idc_auxiliary_dev, adev);
+	adapter = dev_get_drvdata(&adev->dev);
+	pdev = eth_dev->eth_info.eth_shared->pdev;
```
can you move this line inside the !adapter block because thats the only block which uses it

tksingh108 replied:
Do you mean "pdev"?
it is being used below also.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_ethtool.c:
```diff
@@ -131,23 +146,27 @@ static int idpf_set_rxfh(struct net_device *netdev,
 	adapter = vport->adapter;
-
-	if (!idpf_is_cap_ena_all(adapter, IDPF_RSS_CAPS, IDPF_CAP_RSS)) {
+	if (!idpf_is_cap_ena_all(&adapter->dev_info->caps,
```
use a macro like idpf_caps(adapter) to go from adapter to caps pointer. This is necessary for consistency and to guard against future churn due to review comments.

tksingh108 replied:
Okay we can add, but the code looks fine to me.

bhatnagarintel replied:
Whenever you are using 2 or more levels of pointer indirection, it is cleaner and robust to use a macro. It guards against code review comments also because you don't want to end up modifying 15 usages of caps vs one change in a header file.

tksingh108 replied:
Okay, I will add the macro.

tksingh108 replied:
Done!

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_ethtool.c:
```diff
@@ -256,7 +281,7 @@ static int idpf_set_channels(struct net_device *netdev,
 	num_req_rx_q = ch->combined_count + ch->rx_count;
 
-	dev = &vport->adapter->pdev->dev;
+	dev = &eth_shared->pdev->dev;
```
use a macro like idpf_pdev_dev(vport) to go from vport to the device pointer of the PCI device .
This comment applies to all usages of pdev->dev

tksingh108 replied:
I think Madhu changed dev to point to aux device. I need to look at that first. I will incorporate along with that.

tksingh108 replied:
I think this has been addressed as part of pdev changes.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_ethtool.c:
```diff
@@ -387,7 +422,7 @@ static int idpf_set_ringparam(struct net_device *netdev,
 	}
 
-	config_data = &vport->adapter->vport_config[idx]->user_config;
+	config_data = &vport->adapter->vport_config->user_config;
```
IF this expression is occurring frequently : vport->adapter->vport_config->user_config 

then again  I suggest writing a macro and using it consistently.

tksingh108 replied:
Done!

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_ethtool.c:
```diff
@@ -1288,9 +1343,9 @@ static int idpf_set_per_q_coalesce(struct net_device *netdev, u32 q_num,
+	struct idpf_eth_adapter *adapter = idpf_netdev_to_adapter(netdev);
 
-	return adapter->msg_enable;
+	return adapter->dev_info->eth_shared->msg_enable;
```
USe a macro to get from adapter to shared adapter pointer

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,199 @@
+			      struct idpf_eth_idc_event *event);
+};
+
+#define idpf_dev_info_to_adapter(dev_info) \
```
This macro is no longer needed

tksingh108 replied:
done!

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/idpf_main.c:
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: GPL-2.0-only
-/* Copyright (C) 2023 Intel Corporation */
+/* Copyright (C) 2024 Intel Corporation */
```
Copyright year should not be changed.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/idpf_main.c:
```diff
@@ -36,48 +38,24 @@ static void idpf_remove(struct pci_dev *pdev)
+	idpf_eth_idc_deinit_shared(&adapter->eth_shared);
 	pci_set_drvdata(pdev, NULL);
 	kfree(adapter);
+
```
Undo this extra line insertion

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/idpf_main.c:
```diff
@@ -278,4 +248,35 @@ static struct pci_driver idpf_driver = {
+{
+	int err;
+
+	err = idpf_eth_idc_driver_register();
```
The sequence should be reversed in my opinion because it is first a pci driver. So do the pci driver registration and the aux driver registration second.

tksingh108 replied:
I believe this is how the ice driver also does. Also I don't see the issue from test either.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.c:
```diff
@@ -0,0 +1,453 @@
+ */
+static struct idpf_eth_idc_auxiliary_driver *idpf_eth_idc_get_driver(void)
+{
+	return idpf_eth_get_driver();
```
Please use the container method used in OOT PR

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.c:
```diff
@@ -0,0 +1,453 @@
+		auxiliary_device_uninit(&adapter->adevs[i]->adev);
+	}
+
+	kfree(adapter->adevs);
```
There is one race condition here. Suppose all the 'release' functions have not finished executing and we free the adevs here, that would be a problem. Please check how OOT is handling it. 
When I say 'release' function, I assume you are using a aux release function similar to OOT 
aux_dev->adev.dev.release = idpf_aux_device_free;

tksingh108 replied:
I did not see issue in the testing for release. IT appears that the system calls them in sequence.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,380 @@
+	adapter->dev_info->start_probe_post_init = false;
+
+	/* Device's pre-initialization */
+	err = idpf_eth_device_pre_init(adapter);
```
I don't see the probe as divided into pre-init, probe and then post-init.
I am not seeing the value of this kind of division. Its just one probe . The pre-init, if any is in the main driver when it allocates the aux device, core dev info, initializes them and does auxiliary_device_init and 'add'.

tksingh108 replied:
It is done on purpose to keep the vport registration serial. As you know all vport registration in parallel doesn't work.  I think we need some kind of logic to make it happen. And I came up with this.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,380 @@
+ * @eth_shared: Ethernet shared information
+ * @reg_offset: Register offset
+ */
+void __iomem *idpf_eth_get_reg_addr(struct idpf_eth_shared *eth_shared,
```
Given the size of the function (1 line), its a good candidate for an inline function in a header file.

tksingh108 replied:
okay I will move it.

tksingh108 replied:
done!

madhuchittim commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,383 @@
+		.name = "eth",
+		.probe = idpf_eth_probe,
+		.remove = idpf_eth_remove,
+		.shutdown = idpf_eth_shutdown,
```
shutdown is not required?

tksingh108 replied:
I will remove it.

tksingh108 replied:
done!

madhuchittim commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -4,125 +4,153 @@
-	IDPF_SR_RSC_CHANGE,
+#include <linux/auxiliary_bus.h>
+
+struct idpf_adapter;
```
With idpf_dev_info_to_adapter macro gone you would not need the above line?

tksingh108 replied:
It's forward declaration of idpf_adapter structure. It is needed in this file. We cannot include idpf.h, in that case this is the only way to make it happen. I am not sure what are you referring to?

burra006 replied:
agreed to remove the adapter reference

tksingh108 replied:
Done!

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,152 @@
+struct idpf_eth_shared {
+	struct pci_dev *pdev;
+	struct idpf_eth_idc_ops eth_idc_ops;
+	u32 msg_enable;
```
we are using netif_msg_enable to initialize this value. Which means it makes more logical to initi and limit usage in eth aux driver. PCI driver should not make any network interface calls.

burra006 replied:
agreed to place it in aux device adapter...and call it only from aux driver

tksingh108 replied:
Done!

burra006 commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,361 @@
+		idpf_handle_event_link(adapter, event->event_data);
+		break;
+
+	case IDPF_ETH_IDC_EVENT_PROBE_POST_INIT:
```
looks like you are receiving the trigger to register netdev through event handler into the aux driver. Not sure why you need to maintain the start_probe_post_init variable in IDC interface??

tksingh108 replied:
The variable has been moved to eth adapter now.

burra006 commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,361 @@
+		mutex_unlock(&adapter->vport_ctrl_lock);
+		break;
+
+	case IDPF_ETH_IDC_EVENT_DEINIT:
```
Can you please explain when the PCI driver will post this event? not clear why deinit(and no init, which i understand is probe) is a specific use case?? it should be covered in either remove or reset flows....

tksingh108 replied:
It is renamed as "IDPF_ETH_IDC_EVENT_RELEASE_RESOURCES", and workaround/investigation is in progress. If workaround doesn't work we may have to keep it.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.c:
```diff
@@ -0,0 +1,451 @@
+ */
+void idpf_eth_idc_device_init(struct idpf_adapter *adapter)
+{
+	struct device *dev = &adapter->pdev->dev;
```
RCT....

tksingh108 replied:
Done!

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,150 @@
+	struct idpf_eth_shared *eth_shared;
+	enum idpf_vport_type vport_type;
+	struct idpf_eth_idc_auxiliary_dev_caps caps;
+	bool start_post_init;
```
agreed to move start_post_init to eth device adapter

tksingh108 replied:
done!

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,150 @@
+	IDPF_ETH_IDC_EVENT_LINK_CHANGE,
+	IDPF_ETH_IDC_EVENT_RESET_INITIATED,
+	IDPF_ETH_IDC_EVENT_RESET_COMPLETE,
+	IDPF_ETH_IDC_EVENT_RELEASE_RESOURCES,
```
Agreed to take another look to remove this completely....

burra006 replied:
still in progress...

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_virtchnl.h:
```diff
@@ -8,63 +8,105 @@ struct idpf_adapter;
+ *		   the recv_buf _cannot_ be on stack if this is async.
+ * @vc_op: virtchnl op to send
+ */
+struct idpf_vc_xn_params {
```
we should move this struct to idc.h to make idc.h independent and can stand alone  with out any external dependencies....

burra006 replied:
> we should move this struct to idc.h to make idc.h independent and can stand alone with out any external dependencies....

On further look, looks like this struct is depending on other xn structs. So please leave it for now. When we enable libeth all these structs will end up in libeth. Which is by default considered standard for both pci and aux drivers.

tksingh108 replied:
Okay sounds good.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.c:
```diff
@@ -0,0 +1,431 @@
+ */
+void idpf_eth_idc_deinit_shared(struct idpf_eth_shared *eth_shared)
+{
+	eth_shared->hw_addr = NULL;
```
This single line function should be avoided. Please make it NULL in calling function itself.

tksingh108 replied:
Resolved.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_common.h:
```diff
@@ -0,0 +1,254 @@
+ * the event type IDPF_ETH_IDC_EVENT_SINGLE_VPORT is routed to only the
+ * intended port.
+ */
+enum idpf_eth_idc_event_type {
```
This enum is used only by pci driver. So cannot be part of common.h. Move it to idpf.h

tksingh108 replied:
Moved.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_common.h:
```diff
@@ -0,0 +1,254 @@
+ * @__IDPF_PROMISC_MC: Multicast promiscuous mode
+ * @__IDPF_USER_FLAGS_NBITS: Must be last
+ */
+enum idpf_user_flags {
```
Should be used only by aux driver. So not part of common.h. move to idpf_eth.h

tksingh108 replied:
Done!

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_common.h:
```diff
@@ -0,0 +1,254 @@
+ * @IDPF_SR_MTU_CHANGE: Soft reset MTU change
+ * @IDPF_SR_RSC_CHANGE: Soft reset RSC change
+ */
+enum idpf_vport_reset_cause {
```
same as above

tksingh108 replied:
done!

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_common.h:
```diff
@@ -0,0 +1,254 @@
+	u16 index;
+	bool default_vport;
+};
+
```
extra line should be removed

tksingh108 replied:
done!

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_common.h:
```diff
@@ -0,0 +1,254 @@
+	IDPF_OTHER_CAPS		= offsetof(struct virtchnl2_get_capabilities,
+					   other_caps),
+};
+
```
additional line

tksingh108 replied:
done!

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_common.h:
```diff
@@ -0,0 +1,254 @@
+ * @IDPF_QUEUE_DEFAULT_VPORT: Queue for Default vport
+ * @IDPF_QUEUE_DYNAMIC_VPORT: Queue for Dynamic vport
+ */
+enum idpf_queue_type {
```
can be deleted totally. above enum is sufficient

tksingh108 replied:
done!

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_common.h:
```diff
@@ -0,0 +1,254 @@
+#define GETMAXVAL(num_bits)		GENMASK((num_bits) - 1, 0)
+#define IDPF_MAX_BUFQS_PER_RXQ_GRP		2
+
+#define IDPF_TX_MIN_PKT_LEN		17
```
should be part of aux driver headers. maybe idpf_txrx.h

tksingh108 replied:
done!

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_common.h:
```diff
@@ -0,0 +1,254 @@
+#define IDPF_MAX_MBXQ				1
+
+#define GETMAXVAL(num_bits)		GENMASK((num_bits) - 1, 0)
+#define IDPF_MAX_BUFQS_PER_RXQ_GRP		2
```
part of aux driver headers ....

tksingh108 replied:
idpf_virtchnl.c also uses it, so I have to keep it here.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_virtchnl.h:
```diff
@@ -8,63 +8,105 @@ struct idpf_adapter;
 struct idpf_vec_regs;
 struct idpf_vport;
-struct idpf_vport_max_q;
+struct idpf_max_q;
```
This file content is no longer reflecting the virtchnl code. So you may move all this code into eth_common.h and get rid of this file.

Also above forward declaration are not required as they are not used in this file.

tksingh108 replied:
Moved to idpf_eth_common.h.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf.h:
```diff
@@ -717,125 +275,96 @@ static inline void __iomem *idpf_get_reg_addr(struct idpf_adapter *adapter,
 
 u8 idpf_vport_get_hsplit(const struct idpf_vport *vport);
 bool idpf_vport_set_hsplit(const struct idpf_vport *vport, u8 val);
+int idpf_init_dflt_mbx(struct idpf_adapter *adapter);
```
vport hsplit functions above should be part of aux driver??

tksingh108 replied:
Moved to idpf_vport.h

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,145 @@
+			       struct msix_entry *msix_table);
+};
+
+struct idpf_eth_idc_auxiliary_driver;
```
you may not need this forward declaration, as i see you are not referencing until you defined it.

tksingh108 replied:
Removed it.

burra006 commented on drivers/net/ethernet/intel/idpf/idpf_eth_idc.h:
```diff
@@ -0,0 +1,145 @@
+	enum idpf_eth_idc_event_code event_code;
+};
+
+struct idpf_eth_idc_dev_info;
```
See if we can avoid the forward declaration

tksingh108 replied:
We agreed to this one, it is needed as it because of the dependency.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,416 @@
+ * idpf_eth_device_post_init_task - Delayed ethernet post initialization task
+ * @work: work_struct handle to ethernet adapter data
+ */
+static void idpf_eth_device_post_init_task(struct work_struct *work)
```
I don't see why this is a task. There is nothing blocking here. So why spawn a task ?

tksingh108 replied:
Please see this flag "adapter->start_post_init" in this function. It is a blocking factor to re spawn.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.c:
```diff
@@ -0,0 +1,416 @@
+			return -ENOMEM;
+
+		/* Initialize dev_info */
+		adapter->dev_info = &eth_dev->eth_info;
```
assignment to dev_info can be deleted . IT is being done below.

bhatnagarintel commented on drivers/net/ethernet/intel/idpf/eth/idpf_eth.h:
```diff
@@ -0,0 +1,178 @@
+	u32 tx_timeout_count;
+	bool req_tx_splitq;
+	bool req_rx_splitq;
+	u32 *vport_ids;
```
Why is this plural and pointer ? I think there is only one vport_id for one ethernet adapter

walking-machine commented on drivers/net/ethernet/intel/idpf/idpf.h:
```diff
@@ -253,11 +224,12 @@ struct idpf_vc_xn_manager;
  */
 struct idpf_adapter {
+	/* Do not move eth_shared location */
+	struct idpf_eth_shared eth_shared;
```
static_assert(offsetof(...) == 0) can be used here
